File: java/src/test/java/io/github/muthuishere/toolnexus/CompactionTest.java
package io.github.muthuishere.toolnexus;

import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Function;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * Context compaction (SPEC §7F) — C1–C6 (11 checks), driven by the shared cross-language
 * fixture {@code examples/compaction/fixture.json}. Proves: no-op below budget (byte-identical),
 * summarize-above-budget, tool-pair safety, tail preservation, system-prompt preservation, and
 * end-to-end through the SHIPPED {@link LlmClient} loop via the {@code beforeLLM} seam. Hermetic:
 * a localhost mock LLM, no network, no live model.
 */
class CompactionTest {

    // -- fixture ----------------------------------------------------------------

    private final String systemPrompt;
    private final int userTurns;
    private final int maxTokens;
    private final int keepTail;

    @SuppressWarnings("unchecked")
    CompactionTest() throws IOException {
        Path fixture = Path.of(TestFixtures.fixture("compaction/fixture.json"));
        Map<String, Object> f = Json.toMap(Files.readString(fixture));
        Map<String, Object> input = (Map<String, Object>) f.get("input");
        Map<String, Object> generator = (Map<String, Object>) input.get("generator");
        Map<String, Object> options = (Map<String, Object>) f.get("options");
        this.systemPrompt = (String) input.get("systemPrompt");
        this.userTurns = ((Number) generator.get("userTurns")).intValue();
        this.maxTokens = ((Number) options.get("maxTokens")).intValue();
        this.keepTail = ((Number) options.get("keepTail")).intValue();
    }

    /**
     * Build the transcript deterministically per the fixture generator: the systemPrompt, then
     * {@code userTurns} groups of exactly four messages (user / assistant+tool_calls / tool /
     * assistant), padded so it overflows a small budget.
     */
    private List<Object> transcript(int turns) {
        List<Object> m = new ArrayList<>();
        m.add(msg("system", systemPrompt));
        for (int i = 0; i < turns; i++) {
            m.add(msg("user", "question " + i + " " + "pad ".repeat(40)));

            Map<String, Object> call = new LinkedHashMap<>();
            call.put("id", "c" + i);
            call.put("type", "function");
            call.put("function", Map.of("name", "lookup", "arguments", "{}"));
            Map<String, Object> assistant = new LinkedHashMap<>();
            assistant.put("role", "assistant");
            assistant.put("content", null);
            assistant.put("tool_calls", List.of(call));
            m.add(assistant);

            Map<String, Object> tool = new LinkedHashMap<>();
            tool.put("role", "tool");
            tool.put("tool_call_id", "c" + i);
            tool.put("content", "result " + i + " " + "data ".repeat(40));
            m.add(tool);

            m.add(msg("assistant", "answer " + i));
        }
        return m;
    }

    private static Map<String, Object> msg(String role, String content) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("role", role);
        m.put("content", content);
        return m;
    }

    /** The deterministic summarizer of the fixture: "summarized <N> messages". */
    private static final Function<List<Object>, String> SUMMARIZE =
            older -> "summarized " + older.size() + " messages";

    @SuppressWarnings("unchecked")
    private static String role(Object m) {
        return (String) ((Map<String, Object>) m).get("role");
    }

    @SuppressWarnings("unchecked")
    private static Object field(Object m, String k) {
        return ((Map<String, Object>) m).get(k);
    }

    private static LlmClient.BeforeLLMEvent ev(List<Object> messages) {
        return new LlmClient.BeforeLLMEvent(messages, List.of(), "m", 0);
    }

    // -- C1: no-op below budget -------------------------------------------------

    @Test
    void c1_belowBudgetIsByteIdenticalNoOp() {
        var hook = Compaction.compactor(new Compaction.Options().maxTokens(100_000).summarize(SUMMARIZE));
        List<Object> msgs = transcript(3);
        LlmClient.LLMOverride out = hook.apply(ev(msgs));
        assertNull(out, "returns nothing (no override) when under budget");
    }

    // -- C2: compacts above budget ---------------------------------------------

    @Test
    void c2_aboveBudgetSummarizesHeadKeepsTail() {
        List<Object> msgs = transcript(userTurns);
        int before = Compaction.estimateTokens(msgs);
        var hook = Compaction.compactor(new Compaction.Options()
                .maxTokens(maxTokens).keepTail(keepTail).summarize(SUMMARIZE));
        LlmClient.LLMOverride out = hook.apply(ev(msgs));

        assertNotNull(out, "compacts when over budget");
        assertTrue(Compaction.estimateTokens(out.messages()) < before, "result is smaller than the original");
        boolean hasSummary = out.messages().stream()
                .anyMatch(m -> String.valueOf(field(m, "content")).startsWith("[Summary of earlier conversation]"));
        assertTrue(hasSummary, "a summary system message is inserted");
    }

    // -- C3: tool-pair safety ---------------------------------------------------

    @Test
    void c3_toolPairSafetyNoOrphanedToolResult() {
        List<Object> msgs = transcript(userTurns);
        var hook = Compaction.compactor(new Compaction.Options()
                .maxTokens(maxTokens).keepTail(keepTail).summarize(SUMMARIZE));
        List<Object> t = hook.apply(ev(msgs)).messages();

        boolean safe = true;
        for (int i = 0; i < t.size(); i++) {
            if ("tool".equals(role(t.get(i)))) {
                Object id = field(t.get(i), "tool_call_id");
                boolean hasParent = false;
                for (int j = 0; j < i; j++) {
                    if ("assistant".equals(role(t.get(j)))) {
                        Object calls = field(t.get(j), "tool_calls");
                        if (calls instanceof List<?> l) {
                            for (Object c : l) {
                                if (c instanceof Map<?, ?> cm && java.util.Objects.equals(cm.get("id"), id)) {
                                    hasParent = true;
                                }
                            }
                        }
                    }
                }
                if (!hasParent) safe = false;
            }
        }
        assertTrue(safe, "no tool message is orphaned from its tool_calls");

        Object firstNonSystem = t.stream().filter(m -> !"system".equals(role(m))).findFirst().orElse(null);
        assertNotNull(firstNonSystem);
        assertEquals("user", role(firstNonSystem), "the tail begins at a clean user turn");
    }

    // -- C4: system prompt preserved -------------------------------------------

    @Test
    void c4_leadingSystemPromptPreservedVerbatim() {
        List<Object> msgs = transcript(userTurns);
        var hook = Compaction.compactor(new Compaction.Options().maxTokens(maxTokens).summarize(SUMMARIZE));
        List<Object> t = hook.apply(ev(msgs)).messages();
        assertEquals("system", role(t.get(0)), "first message is still system");
        assertTrue(String.valueOf(field(t.get(0), "content")).contains("SOUL"),
                "original SOUL system prompt is still first, verbatim");
    }

    // -- C5: flush-to-memory nudge ---------------------------------------------

    @Test
    void c5_flushToMemoryInjectsReminder() {
        List<Object> msgs = transcript(userTurns);
        var hook = Compaction.compactor(new Compaction.Options()
                .maxTokens(maxTokens).summarize(SUMMARIZE).flushToMemory(true));
        List<Object> t = hook.apply(ev(msgs)).messages();
        boolean hasReminder = t.stream()
                .anyMatch(m -> String.valueOf(field(m, "content")).contains("save it with the memory tool"));
        assertTrue(hasReminder, "a 'save with the memory tool' reminder is present");
    }

    // -- C6: end-to-end through the SHIPPED client loop -------------------------

    @Test
    void c6_wiredViaBeforeLLMLoopKeepsGoingPastCompaction() throws IOException {
        // Mock LLM: makes 6 tool calls (padding the transcript), then a final answer that reports
        // whether it saw a compacted transcript ([Summary ...]).
        AtomicInteger calls = new AtomicInteger(0);
        HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/", exchange -> {
            String body = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
            Map<String, Object> req = Json.toMap(body);
            @SuppressWarnings("unchecked")
            List<Object> messages = (List<Object>) req.get("messages");
            boolean compacted = messages != null && messages.stream()
                    .anyMatch(m -> String.valueOf(field(m, "content")).startsWith("[Summary"));
            int c = calls.get();
            boolean done = c >= 6;
            Map<String, Object> message;
            if (done) {
                Map<String, Object> mm = new LinkedHashMap<>();
                mm.put("role", "assistant");
                mm.put("content", "final (compacted=" + compacted + ")");
                message = mm;
            } else {
                calls.incrementAndGet();
                Map<String, Object> call = new LinkedHashMap<>();
                call.put("id", "c" + c);
                call.put("type", "function");
                call.put("function", Map.of("name", "pad", "arguments", "{}"));
                Map<String, Object> mm = new LinkedHashMap<>();
                mm.put("role", "assistant");
                mm.put("content", null);
                mm.put("tool_calls", List.of(call));
                message = mm;
            }
            Map<String, Object> data = Map.of(
                    "choices", List.of(Map.of("message", message)),
                    "usage", Map.of("prompt_tokens", 50, "completion_tokens", 10, "total_tokens", 60));
            respond(exchange, 200, Json.stringify(data));
        });
        server.start();
        String base = "http://127.0.0.1:" + server.getAddress().getPort();

        Tool pad = NativeTool.of("pad", "pads", null, args -> "x ".repeat(600));
        try (Toolkit tk = Toolkit.create(new Toolkit.Options().builtins(false).extraTools(pad))) {
            LlmClient.Hooks hooks = new LlmClient.Hooks().beforeLLM(
                    Compaction.compactor(new Compaction.Options().maxTokens(600).keepTail(250).summarize(SUMMARIZE)));
            LlmClient client = LlmClient.create(new LlmClient.Options()
                    .baseUrl(base).style("openai").model("m").apiKey("spike").maxTurns(20)
                    .systemPrompt(systemPrompt).hooks(hooks));

            LlmClient.RunResult r = client.run("start", tk);
            assertEquals("done", r.status, "run completes to a final answer despite mid-run compaction");
            assertTrue(r.text.startsWith("final"), "final answer reached: " + r.text);
            assertTrue(r.text.contains("compacted=true"), "compaction actually fired during the run: " + r.text);
            assertTrue(Compaction.estimateTokens(r.messages) < 4000,
                    "final transcript is bounded, not full raw history: " + Compaction.estimateTokens(r.messages));
        } finally {
            server.stop(0);
        }
    }

    private static void respond(HttpExchange exchange, int status, String body) throws IOException {
        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().set("Content-Type", "application/json");
        exchange.sendResponseHeaders(status, bytes.length);
        exchange.getResponseBody().write(bytes);
        exchange.close();
    }
}

File: elixir/test/agents/compaction_test.exs
defmodule Toolnexus.Agents.CompactionTest do
  @moduledoc """
  Context compaction (SPEC §7F) — the C1–C6 acceptance scenarios (11 checks), ported
  from `js/spike/compaction-demo.ts` and asserted against the shared
  `examples/compaction/fixture.json` transcript shape:

    * C1 — no-op below budget (byte-identical)
    * C2 — summarize head, keep tail, above budget
    * C3 — tool-pair safety: no orphaned tool result; tail begins at a user turn
    * C4 — leading system prompt preserved verbatim
    * C5 — flush_to_memory injects a pre-compact reminder
    * C6 — wired via before_llm in a real run through the shipped client loop

  The transcript is generated deterministically from the fixture (system prompt +
  `userTurns` groups of user / assistant+tool_calls / tool / assistant), so this port
  produces the same bytes as every other.
  """
  use ExUnit.Case, async: true

  alias Toolnexus.Agents.Compaction
  alias Toolnexus.{Client, Native}

  @repo_root Path.expand("../../..", __DIR__)
  @fixture Path.join(@repo_root, "examples/compaction/fixture.json")

  # Build the transcript exactly as examples/compaction/fixture.json prescribes.
  defp transcript(user_turns) do
    fix = @fixture |> File.read!() |> Jason.decode!()
    system_prompt = fix["input"]["systemPrompt"]
    pad = String.duplicate("pad ", 40)
    data = String.duplicate("data ", 40)

    groups =
      Enum.flat_map(0..(user_turns - 1), fn i ->
        [
          %{"role" => "user", "content" => "question #{i} #{pad}"},
          %{
            "role" => "assistant",
            "content" => nil,
            "tool_calls" => [
              %{"id" => "c#{i}", "type" => "function", "function" => %{"name" => "lookup", "arguments" => "{}"}}
            ]
          },
          %{"role" => "tool", "tool_call_id" => "c#{i}", "content" => "result #{i} #{data}"},
          %{"role" => "assistant", "content" => "answer #{i}"}
        ]
      end)

    [%{"role" => "system", "content" => system_prompt} | groups]
  end

  defp summarize(older), do: "summarized #{length(older)} messages"

  defp ev(messages), do: %{messages: messages, tools: [], model: "m", turn: 0}

  # C1 -------------------------------------------------------------------------
  test "C1: below budget → no-op (byte-identical)" do
    hook = Compaction.compactor(max_tokens: 100_000, summarize: &summarize/1)
    assert hook.(ev(transcript(3))) == nil
  end

  # C2 -------------------------------------------------------------------------
  test "C2: above budget → summarize head, keep tail" do
    msgs = transcript(30)
    before = Compaction.estimate_tokens(msgs)
    hook = Compaction.compactor(max_tokens: 2000, keep_tail: 800, summarize: &summarize/1)
    out = hook.(ev(msgs))

    assert %{messages: result} = out
    assert Compaction.estimate_tokens(result) < before

    assert Enum.any?(result, fn m ->
             String.starts_with?(to_string(m["content"]), "[Summary of earlier conversation]")
           end)
  end

  # C3 -------------------------------------------------------------------------
  test "C3: tool-pair safety — no orphaned tool result; tail begins at a user turn" do
    hook = Compaction.compactor(max_tokens: 2000, keep_tail: 800, summarize: &summarize/1)
    %{messages: t} = hook.(ev(transcript(30)))

    safe? =
      t
      |> Enum.with_index()
      |> Enum.all?(fn
        {%{"role" => "tool", "tool_call_id" => id}, i} ->
          t
          |> Enum.take(i)
          |> Enum.any?(fn m ->
            m["role"] == "assistant" and
              Enum.any?(m["tool_calls"] || [], &(&1["id"] == id))
          end)

        _ ->
          true
      end)

    assert safe?, "a tool message was orphaned from its tool_calls"

    first_non_system = Enum.find(t, &(&1["role"] != "system"))
    assert first_non_system["role"] == "user"
  end

  # C4 -------------------------------------------------------------------------
  test "C4: leading system prompt preserved verbatim" do
    hook = Compaction.compactor(max_tokens: 2000, summarize: &summarize/1)
    %{messages: [first | _]} = hook.(ev(transcript(30)))
    assert first["role"] == "system"
    assert String.contains?(first["content"], "SOUL")
  end

  # C5 -------------------------------------------------------------------------
  test "C5: flush_to_memory injects a pre-compact reminder" do
    hook = Compaction.compactor(max_tokens: 2000, summarize: &summarize/1, flush_to_memory: true)
    %{messages: result} = hook.(ev(transcript(30)))

    assert Enum.any?(result, fn m ->
             String.contains?(to_string(m["content"]), "save it with the memory tool")
           end)
  end

  # --- coverage: edge paths ---------------------------------------------------

  test "keep_tail too small for any group → falls back to the most recent user turn (safety over size)" do
    msgs = transcript(30)
    # keep_tail = 1 fits no user-boundary tail, so the split rule falls back to the
    # last user turn; a tool result must still never be orphaned.
    hook = Compaction.compactor(max_tokens: 2000, keep_tail: 1, summarize: &summarize/1)
    %{messages: t} = hook.(ev(msgs))

    # first non-system message is a user turn
    assert Enum.find(t, &(&1["role"] != "system"))["role"] == "user"
    # summary present, and the older body was actually compacted (result smaller)
    assert Enum.any?(t, &String.starts_with?(to_string(&1["content"]), "[Summary"))
    assert Compaction.estimate_tokens(t) < Compaction.estimate_tokens(msgs)
  end

  test "nothing safely compactible above the preserved head → no-op (nil)" do
    # A transcript that is only the preserved system head: there is no conversational
    # body to summarize, so even 'over budget' the hook is a no-op.
    msgs = [%{"role" => "system", "content" => "You are Ava. SOUL."}]

    hook = Compaction.compactor(max_tokens: 0, keep_tail: 100_000, summarize: &summarize/1)
    assert hook.(ev(msgs)) == nil
  end

  test "count_tokens override replaces the default estimator" do
    msgs = transcript(3)
    # Force 'over budget' via a custom counter so compaction fires on a small transcript.
    hook =
      Compaction.compactor(
        max_tokens: 5,
        keep_tail: 2,
        summarize: &summarize/1,
        count_tokens: fn m -> length(m) end
      )

    assert %{messages: result} = hook.(ev(msgs))
    assert Enum.any?(result, &String.starts_with?(to_string(&1["content"]), "[Summary"))
  end

  test "estimate_tokens sums ceil(byte_size(json)/4) per message" do
    m = %{"role" => "user", "content" => "hello"}
    # Jason.encode! is compact, no HTML escaping, no trailing newline — byte-length parity with JS.
    expected = ceil(byte_size(Jason.encode!(m)) / 4)
    assert Compaction.estimate_tokens([m]) == expected
  end

  # C6 -------------------------------------------------------------------------
  test "C6: wired via before_llm — the loop keeps going past a compaction" do
    {:ok, calls} = Agent.start_link(fn -> 0 end)

    transport = fn req ->
      n = Agent.get_and_update(calls, &{&1, &1 + 1})

      compacted =
        Enum.any?(req.body["messages"], fn m ->
          String.starts_with?(to_string(m["content"] || ""), "[Summary")
        end)

      message =
        if n >= 6 do
          %{"role" => "assistant", "content" => "final (compacted=#{compacted})"}
        else
          %{
            "role" => "assistant",
            "content" => nil,
            "tool_calls" => [
              %{"id" => "c#{n}", "type" => "function", "function" => %{"name" => "pad", "arguments" => "{}"}}
            ]
          }
        end

      body = %{
        "choices" => [%{"message" => message}],
        "usage" => %{"prompt_tokens" => 50, "completion_tokens" => 10, "total_tokens" => 60}
      }

      {:ok, %{status: 200, headers: %{}, body: body}}
    end

    pad =
      Native.define_tool(
        name: "pad",
        description: "pads",
        execute: fn _args -> String.duplicate("x ", 600) end
      )

    client =
      Client.create(
        base_url: "http://mock.local",
        style: "openai",
        model: "m",
        api_key: "spike",
        max_turns: 20,
        transport: transport,
        system_prompt: "You are Ava. SOUL.",
        hooks: %{before_llm: Compaction.compactor(max_tokens: 600, keep_tail: 250, summarize: &summarize/1)}
      )

    r = Client.run(client, "start", [pad])

    assert r.status == "done" and String.starts_with?(r.text, "final"), r.text
    assert String.contains?(r.text, "compacted=true"), r.text
    assert Compaction.estimate_tokens(r.messages) < 4000
  end
end
